include/swift/Runtime/Metadata.h

```diff
@@ -2642,24 +2647,20 @@ swift_getFunctionTypeMetadata(const void *flagsArgsAndResult[]);
 
 SWIFT_RUNTIME_EXPORT
 const FunctionTypeMetadata *
-swift_getFunctionTypeMetadata1(FunctionTypeFlags flags,
-                               const void *arg0,
+swift_getFunctionTypeMetadata1(FunctionTypeFlags flags, const void *arg0,
+                               const void *paramFlags,
```

Comment:
If we're no longer mangling things into the low bits of the parameter types, we can just use 'const Metadata *' instead of 'void *', since we're not trying to encourage caution.

Why is paramFlags a void*?

getFunctionTypeMetadataN is a convenience function for common cases.  Do we have any reason to think that parameter type flags are a common enough case to justify making these functions less convenient?

You also need to update all the getFooFunctionTypeMetadata functions below this, or at least their most-general cases.

Reply:
I wasn’t sure what to you for parameter flags here, should it be `uint32_t *` instead? I think parameter flags for single parameter are pretty common, I’ve seen examples in compatibility suite.

Reply:
I don't doubt that they occur, just that they occur in enough frequency to warrant adding extra requirements to the common case.  Can you get counts here?

Reply:
I've found 87 occurrences of functions with `inout` and 1-3 arguments in source compatibility projects, most of them in `GRDB` project. I'm pretty sure there is going to be much more with `__shared` when that's available... I'm also curious - if not array of parameters how else should we track flags in this case, with individual arguments?

Reply:
Do you happen to have the full histogram for the source compatibility suite?  Specifically I'd like to know the cross table of (convention) x (argument count) x (has any flags).

I agree that __shared/__owned will make flags more common.  (We'll have to decide which is the default, and obviously it should track the language decision because the language decision will make the default much more common than the other.)

Our original decision to go up to 3 was not based on any particular insight that I remember.  The purpose of these functions is to optimize code size for common cases, and setting up a call with 7 arguments is already a ton of code, so the benefit might be negligible.  Maybe we'll want to provide specializations for (1,false), (2, false), (3, false) but just (1, true).  Certainly it seems questionable whether we should do these common-path entrypoints for each of the secondary ``@convention``s; maybe we should just have common-path entrypoints that take the convention as an argument.  Anyway, data is the key here.

Reply:
Unfortunately no, I just made a grep over all of the projects in the `project_cache` folder, but I will try to get all of the info with some effort.

> Maybe we'll want to provide specializations for (1,false), (2, false), (3, false) but just (1, true)...

You mean like `get_swiftFunctionTypeMetadata1InOut`, `get_swiftFunctionTypeMetadata1Shared` etc. ?

I'm actually wondering how many more flags are there possibly we'd want to add? It seems like from `MetadataCache` perspective it would be much easier to just make a custom metadata type ref for parameters with flags, which would encode flags as part of type metadata body but I'm not sure if that's a good idea and how feasible that would be? It seems at least it would be much easier to unique - no flags - pointer to metadata of type, at least one flag - pointer to metadata ref for "parameter" with includes metadata ref for type + uint32_t for flags, WDYT?

Reply:
Oh, are you just grepping for typealiases of function type or something?  We don't actually need to build type metadata for every function type we see.  Although I guess it's a significantly larger population, and I have no reason to think it's likely to be unrepresentative of the population of function types that we need metadata for.

RE: (1, false), I meant something more like swift_getFunctionTypeMetadataWithFlags1 vs. something that just assumed the default flags.

RE: MetadataCache: if MetadataCache makes it awkward to handle heterogenous key data, we can easily just drop down to SimpleGlobalCache (i.e. ConcurrentMap) the same way that several of the other caches do.  MetadataCache is primarily there to provide additional locking so that we can ensure that the more expensive metadata kinds are uniquely initialized, and function-type metadata don't need that.

Reply:
> Oh, are you just grepping for typealiases of function type or something?

Yes, that's exactly what I was doing, I wish we had some tools to generate reports like that but I'm afraid we don't, or maybe I just don't know about them. My thinking was that if there aren't many functions with `inout` (overall) in the project then and even fewer are going to require metadata generation that says something about distribution...

RE: `swift_getFunctionTypeMetadataWithFlags1` - Sure! I got that idea as well, just wasn't sure if you are proposing something even more sophisticated... I think it makes sense because many 1/2/3 parameter functions are likely not to have any flags, judging by compatibility suite. Which brings us to question about having separate "function parameter metadata" which is going to include pointer to type + uint32_t for flags and not going to require any changes to `swift_getFunctionTypeMetadata*` functions and `MetadataCache`, WDYT?

RE: MetadataCache - Ah, I see, that make sense to me!

Reply:
I think having a separate metadata object for each parameter would be rather gratuitously wasteful, and I can't think of anything that would benefit from it.

Reply:
Not always but only if there are flags present, if that’s not a very common occurance it’s not going to waste too much space, right?

Reply:
I think there's some value in having the layout of the component types not vary that drastically based on whether flags are present.  This is why I was suggesting having separate arrays.

Reply:
I totally agree! The reason why I bring up this parameter type ref metadata idea is because it makes it much easier to work with function metadata and doesn't require any specific optimizations, like in case of the array of flags, we'd like to keep only flags for parameters which have them, and don't use any additional place, such most likely would be more complicated comparing to extracting a single `uint32_t` from metadata ref and this approach wouldn't require adding even more optimized calls for 1/2/3 parameter functions with and w/e flags...